lib/vm/src/trap/traphandlers.rs

```diff
@@ -547,106 +572,104 @@ where
         Self {
-            unwind: Cell::new(UnwindReason::None),
-            vmctx,
+            unwind: UnsafeCell::new(MaybeUninit::uninit()),
             jmp_buf: Cell::new(ptr::null()),
             reset_guard_page: Cell::new(false),
-            prev: None,
+            prev: Cell::new(ptr::null()),
+            trap_handler,
             handling_trap: Cell::new(false),
         }
     }
 
-    fn with(mut self, closure: impl FnOnce(&Self) -> i32) -> Result<(), Trap> {
-        tls::with(|prev| {
-            self.prev = prev.map(|p| p as *const _);
-            let ret = tls::set(&self, || closure(&self));
-            match self.unwind.replace(UnwindReason::None) {
-                UnwindReason::None => {
-                    debug_assert_eq!(ret, 1);
-                    Ok(())
-                }
-                UnwindReason::UserTrap(data) => {
-                    debug_assert_eq!(ret, 0);
-                    Err(Trap::new_from_user(data))
-                }
-                UnwindReason::LibTrap(trap) => Err(trap),
-                UnwindReason::RuntimeTrap {
-                    backtrace,
-                    pc,
-                    signal_trap,
-                } => {
-                    debug_assert_eq!(ret, 0);
-                    Err(Trap::new_from_wasm(pc, backtrace, signal_trap))
-                }
-                UnwindReason::Panic(panic) => {
-                    debug_assert_eq!(ret, 0);
-                    std::panic::resume_unwind(panic)
-                }
+    fn with(self, closure: impl FnOnce(&CallThreadState) -> i32) -> Result<(), Trap> {
+        let ret = tls::set(&self, || closure(&self))?;
+        if ret != 0 {
+            return Ok(());
+        }
+        match unsafe { (*self.unwind.get()).as_ptr().read() } {
+            UnwindReason::UserTrap(data) => {
+                debug_assert_eq!(ret, 0);
+                Err(Trap::User(data))
             }
-        })
-    }
-
-    fn any_instance(&self, func: impl Fn(&Instance) -> bool) -> bool {
-        unsafe {
-            if func(
-                self.vmctx
-                    .vmctx
-                    .as_ref()
-                    .expect("`VMContext` is null in `any_instance`")
-                    .instance(),
-            ) {
-                return true;
+            UnwindReason::LibTrap(trap) => Err(trap),
+            UnwindReason::WasmTrap {
+                backtrace,
+                pc,
+                signal_trap,
+            } => {
+                debug_assert_eq!(ret, 0);
+                Err(Trap::wasm(pc, backtrace, signal_trap))
             }
-            match self.prev {
-                Some(prev) => (*prev).any_instance(func),
-                None => false,
+            UnwindReason::Panic(panic) => {
+                debug_assert_eq!(ret, 0);
+                std::panic::resume_unwind(panic)
             }
         }
     }
 
     fn unwind_with(&self, reason: UnwindReason) -> ! {
```

Comment:
Shouldn't this be `mut self` since this is a mutating function?

Reply:
I tried to deviate minimally on this from prev implementation on this. I tried it after your comment though, and it is technically impossible because of this:
```
   --> lib/vm/src/trap/traphandlers.rs:644:39
    |
644 |         let ret = tls::set(&mut self, || closure(&mut self))?;
    |                   -------- ---------  ^^              ---- second borrow occurs due to use of `self` in closure
    |                   |        |          |
    |                   |        |          second mutable borrow occurs here
    |                   |        first mutable borrow occurs here
    |                   first borrow later used by call
```